### ISSUE TITLE
`ak.to_categorical()` does not actually store unique strings

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#2129]
### Version of Awkward Array

main

### Description and code to reproduce

In this example one can see that the memory consumption of the array is linear in the number of items, even though they're repeated
```python
>>> ak.to_categorical(["thisisaverylongstring!!!!!!!thisisaverylongstring!!!!!!!"]).nbytes
80
>>> ak.to_categorical(["thisisaverylongstring!!!!!!!thisisaverylongstring!!!!!!!", "thisisaverylongstring!!!!!!!thisisaverylongstring!!!!!!!"]).nbytes
144
```

Perhaps caused by this, or just not implemented; merging two categoricals merges the contents naively.

**Comment by jpivarski** [comment]
Not quite a bug because it was intended to work this way, but I can see the value in making the `ak.to_categorical` function be more active. We do have an L2 or L3 function for finding unique strings.

(`nbytes` can be a little deceptive because it naively ignores overlaps, following NumPy, which ignores strides and the size of the `base` array when reporting `nbytes`. But I checked this example by looking at the layout.)

**Comment by agoose77** [comment]
> (nbytes can be a little deceptive because it naively ignores overlaps

Yep, I'm being lazy here — if the array were truly categorical, then it should grow non-linearly with additional values in the same categories. However, I assume things like contiguousness etc that make it a poor metric. 

In any case, we should be able to vectorise this down the road, i.e. write a kernel for hashing the known awkward types. However, for now, that's a todo.

> Not quite a bug because it was intended to work this way, 

I don't think we do want this to be the case — this means that `ak.to_categorical()` doesn't produce a useful output for string types, right? i.e., it saves no memory. I might be missing something, though.

**Comment by agoose77** [comment]
Fixed by #2779 (or at least, not relevant!)
